rs/replicated_state/src/canister_state/queues/tests.rs

```diff
@@ -2218,6 +2209,41 @@ fn output_into_iter_peek_with_stale_references(
     prop_assert_eq!(None, output_iter.next());
 }
 
+#[test_strategy::proptest]
+fn output_into_iter_pop_with_stale_references(
+    #[strategy(arb_canister_output_queues(10, Some(5)))] test: (
+        CanisterQueues,
+        VecDeque<RequestOrResponse>,
+    ),
+    #[any] deadline: u32,
+) {
+    let (mut canister_queues, _raw_requests) = test;
+    let own_canister_id = canister_test_id(13);
+    let local_canisters = BTreeMap::new();
+    // Time out some messages.
+    canister_queues.time_out_requests(
+        coarse_time(deadline).into(),
+        &own_canister_id,
+        &local_canisters,
+    );
+
+    // Pop (after optionally peeking) a few times.
+    let mut output_iter = canister_queues.output_into_iter();
+    let mut should_peek = deadline % 2 == 0;
+    for _ in 0..3 {
+        if should_peek {
+            output_iter.peek();
+        }
+        if output_iter.next().is_none() {
+            break;
+        };
+        should_peek = !should_peek;
+    }
+
+    // Invariants should hold.
+    prop_assert!(canister_queues.stats_ok());
```

Comment:
Is this checking all the relevant invariants or should we check additional things?

Reply:
In #879 it is replaced by `canister_queues.test_invariants()`, which is this plus what used to be `queues_ok()`.

This is a test I brought back from my attempt to drop the "front of queue not stale" invariant. The test was intended to make sure we no longer enforced that invariant (i.e. that the "queues were OK" even with a stale reference at the front). Right now, it tests that popping only some messages from an output queue with stale references leaves it in a consistent state (which is what `test_invariants()` will do).